internal/hcsoci/network.go

```diff
@@ -68,3 +69,88 @@ func GetNamespaceEndpoints(ctx context.Context, netNS string) ([]*hns.HNSEndpoin
 	}
 	return endpoints, nil
 }
+
+// Network namespace setup is a bit different for templates and clones.
+// A normal network namespace creation works as follows: When a new Pod is created, HNS is
+// called to create a new namespace and endpoints inside that namespace. Then we hot add
+// this network namespace and then the network endpoints associated with that namespace
+// into this pod UVM. Later when we create containers inside that pod they start running
+// inside this namespace. The information about namespace and endpoints is maintained by
+// the HNS and it is stored in the guest VM's registry (UVM) as well. So if inside the pod
+// we see a namespcae with id 'NSID' then we can query HNS to find all the information
+// about that namespace.
+//
+// When we clone a VM (with containers running inside it) we can hot add a new namespace
+// and endpoints created for that namespace to the uvm but we can't make the existing
+// processes/containers to automatically switch over to this new namespace. This means
+// that all clones will have the same namespace ID as that of the template from which they
+// are created. This can make debugging very difficult when we have multiple UVMs with
+// same NSID.  To make debugging a bit easier we will create all templates and clones with
+// a special NSID that is dedicated for cloning. So during debugging if we see multiple
+// UVMs with same NSID we can quickly check if this is because of cloning. Now inside the
+// HNS you can not have multiple UVMs using the same NSID. So to achieve this when we
+// create a template or a cloned pod we will ask HNS to create a new namespace and
+// endpoints for that UVM but when we actually send a request to hot add that namespace we
+// will change the namespace ID with the ID that is specifically created for cloning
+// purposes. Similarly, when hot adding an endpoint we will modify this endpoint
+// information to set its network namespace ID to this default ID. This way inside every
+// template and cloned pod the namespace ID will remain same (but each cloned UVM will
+// have a different endpoints) but the HNS will have the actual namespace ID that was
+// created for that UVM.
```

Comment:
This comment is really helpful, but I feel like it might be too long for keeping it here. I think ideally we should just mention what is absolutely necessary here and then if we want to have more details about it, put it in a doc. Thoughts?

Reply:
I agree that it is a big too big. I will move this to the doc and just add a reference here for the doc.